src/process_utility.c

```diff
@@ -1200,18 +1204,94 @@ static void
 process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
 {
 	CreateIndexInfo *info = (CreateIndexInfo *) arg;
+
 	IndexStmt  *stmt = transformIndexStmt(chunk_relid, info->stmt, NULL);
 	Chunk	   *chunk = ts_chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
 
 	ts_chunk_index_create_from_stmt(stmt, chunk->fd.id, chunk_relid, ht->fd.id, info->obj.objectId);
 }
 
+static void
+process_index_chunk_multitransaction(Hypertable *ht, Oid chunk_relid, void *arg)
+{
+	CreateIndexInfo *info = (CreateIndexInfo *) arg;
+	Chunk	   *chunk;
+	Relation	hypertable_rel;
+	Relation	hypertable_index_rel;
+	Relation	chunk_rel;
+	IndexStmt  *stmt;
+	CatalogSecurityContext sec_ctx;
+
+	/*
+	 * commit root table's transaction if it exists (foreach_chunk requires a
+	 * transcation running, so we cannot commit before this point)
+	 */
+	if (ActiveSnapshotSet())
+		PopActiveSnapshot();
+
+	if (IsTransactionOrTransactionBlock())
+		CommitTransactionCommand();
+
+	/* Start a new transaction for each relation. */
+	StartTransactionCommand();
+	/* functions in indexes may want a snapshot set */
+	Snapshot	snapshot = GetTransactionSnapshot();
+
+	PushActiveSnapshot(snapshot);
+
+	/*
+	 * Change user since chunk's are typically located in an internal schema
+	 * and chunk indexes require metadata changes. In regular  CREATE INDEX we
+	 * do this in process_index_end
+	 */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	/*
+	 * Hold a lock on the hypertable, it's index index, and the chunk to
+	 * prevent from being altered. Since we use the same relids across
+	 * transactions, there is a potential issue if the id gets reassigned
+	 * between one sub-transaction and the next. CLUSTER has a similar issue.
+	 */
+	hypertable_rel = relation_open(ht->main_table_relid, AccessShareLock);
```

Comment:
@davidkohn88 That was supposed to be a `ShareLock` the `AccessShareLock` was supposed to go on the chunk.  
I have to admit I'm not entirely sure what lock is needed here, I _think_ the lock on the base index suffices to prevent it from being altered while this transaction is in progress, but I'm not comfortable doing only that unless I'm really sure.

Reply:
Right. I don't know that we need a lock on the hypertable itself at all.